src/MongoDB.Driver.Core/Core/Compression/Snappy/SnappyCodec.cs

```diff
@@ -0,0 +1,168 @@
+                    throw new InvalidDataException("Input is not a valid snappy-compressed block.");
+            }
+        }
+
+        public static int GetUncompressedLength(byte[] input)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+
+            return GetUncompressedLength(input, 0, input.Length);
+        }
+
+        public static int Uncompress(byte[] input, int inputOffset, int inputLength, byte[] output, int outputOffset, int outputLength)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+            Ensure.IsNotNull(output, nameof(output));
+            EnsureBufferRangeIsValid(inputOffset, inputLength, input.Length);
+            EnsureBufferRangeIsValid(outputOffset, outputLength, output.Length);
+
+            if (inputLength == 0)
+            {
+                throw new InvalidDataException("Compressed block cannot be empty.");
+            }
+
+            if (outputOffset == output.Length)
+            {
+                output = new byte[1];
+                outputOffset = 0;
+            }
+
+            var status = SnappyAdapter.snappy_uncompress(input, inputOffset, inputLength, output, outputOffset, ref outputLength);
+            switch (status)
+            {
+                case SnappyStatus.Ok:
+                    return outputLength;
+                case SnappyStatus.BufferTooSmall:
+                    throw new ArgumentOutOfRangeException("Output array is too small.");
+                default:
+                    throw new InvalidDataException("Input is not a valid snappy-compressed block.");
+            }
+        }
+
+        public static byte[] Uncompress(byte[] input)
+        {
+            var max = GetUncompressedLength(input);
+            var output = new byte[max];
+            var outputLength = Uncompress(input, 0, input.Length, output, 0, output.Length);
+            if (outputLength == max)
+            {
+                return output;
+            }
+
+            var truncated = new byte[outputLength];
+            Array.Copy(output, truncated, outputLength);
+            return truncated;
+        }
+
+        public static bool Validate(byte[] input, int inputOffset, int inputLength)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+            EnsureBufferRangeIsValid(inputOffset, inputLength, input.Length);
+            if (inputLength == 0)
+            {
+                return false;
+            }
+
+            return SnappyAdapter.snappy_validate_compressed_buffer(input, inputOffset, inputLength) == SnappyStatus.Ok;
+        }
+
+        public static bool Validate(byte[] input)
+        {
+            Ensure.IsNotNull(input, nameof(input));
+
+            return Validate(input, 0, input.Length);
+        }
+
+        // private static methods
+        private static void EnsureBufferRangeIsValid(int offset, int length, int bufferLength)
+        {
+            if (offset < 0 || length < 0 || offset + length > bufferLength)
```

Comment:
I think it should be `length <= 0`.

I see indications elsewhere that zero length buffers are not valid.

Reply:
Discussed in the slack. Since sometimes empty buffers are valid, this change is not required.